subsys/power/power.c

```diff
@@ -116,6 +117,17 @@ enum power_states _sys_suspend(s32_t ticks)
 		 * in deep sleep mode.
 		 */
 		_sys_pm_idle_exit_notification_disable();
+#if CONFIG_DEVICE_POWER_MANAGEMENT
+	} else if (sys_pm_policy_low_power_devices(pm_state)) {
+		/* low power peripherals. */
+		if (sys_pm_low_power_devices()) {
```

Comment:
@wentongwu  My two cents... I started looking at power management recently, and based on what I have seen thus far, I tend to agree with the comments from @pizi-nordic. My general impression is that the centralized method isn't very useful by itself for the reasons @pizi-nordic stated. It is trying to couple device power states with system power states, which I don't think is the typical scenario.

For optimal power management, we want to keep devices in their low-power state for as long as possible, not just when the CPU goes to sleep/deep-sleep. While it is useful for a device to know about the CPU's power transitions (e.g. for saving/restoring state that might be impacted), it is rarely the determining factor in changing a device's power state.

I see the "centralized method" more as a way to notify a device of system power states, so that it gets a chance to save any context if necessary, rather than initiating an actual device state change. What do you think of replacing calls to sys_pm_suspend_devices() and sys_pm_low_power_devices() with 'sys_pm_notify_device(pm_state)', which would be more in tune with how I think it should work?

Other more general thoughts:

- If the device is busy doing something that requires the CPU being active, my understanding is that it should inform the power manager with a call to device_busy_set(). It'd be nicer if there is a form of this API that passes in the system power state the device cares about (e.g. device_busy_set(DISALLOW_SYS_POWER_STATE_DEEP_SLEEP_2)). The device might want a way to say "I am too busy to deal with the CPU entering deep sleep #2, but sleep #1 is ok"). The suspend logic should ensure that it doesn't perform a CPU power transition a device doesn't wish to occur. It is unclear to me whether the suspend logic is taking busy states into consideration today.

- One other thing that seems missing is power management at the power domain level. It'd be great to automatically turn off unused power domains on the CPU while it is in active state, and have a way for the devices declare their dependencies on these. In this manner, if none of the devices used in an application depends on a power domain, it can safely be shut off.

Reply:
@vanti @pizi-nordic Thanks for the review and sorry for the late response. I mostly agree with your idea and also the power domain consideration is also in the plan. I will firstly add the notification mechanism to have devices know CPU's power state. The return value will be the indication whether devices is happy with that state or not, but doesn't consider device's suggestion currently. And I will keep the "centralized method" in case anyone need it, like the power button and in Android system. How do you think of that? Thanks

Reply:
I didn't read the discussion in details, but I see interest in some use cases for "centralized method". This could be used, for some architectures lacking pm features (like LP timer), by application to manage a pm scenario. Application can decide to enter the low power state with `sys_suspend` after it has configured proper ways to wake up (like RTC alarm, GPIOs). So I'm good with keeping the "centralized method".

Reply:
In my oppinion, the centralized control is something which should be implemented layer above. So instead of linking System power management to device power management we should have a mechanism which order both to reach the equivalent power state at the same time.

Reply:
@wentongwu I am fine with keeping the "centralized method" if there are use-cases for it. My concern is that looking at the code it seems to be the only supported method today when CONFIG_DEVICE_POWER_MANAGEMENT=y is set, which gives the impression that device power state transitions are supposed to be initiated by system power state transitions. So yes having a notification mechanism and a way to turn off the "centralized method" would be helpful.

Reply:
@pizi-nordic @vanti Could you please give some comments for the developing patch first? And it's about the notification mechanism, centralized method will be considered more. Thanks.